presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.log.Logger;
+import io.prestosql.execution.TaskId;
+import io.prestosql.spi.QueryId;
+import io.prestosql.spi.predicate.TupleDomain;
+
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.prestosql.spi.predicate.TupleDomain.columnWiseUnion;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Predicate.isEqual;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final ConcurrentHashMap<SourceDescriptor, DynamicFilterSummaryWithSenders> dynamicFilterSummaries = new ConcurrentHashMap<>();
+    private static final Logger log = Logger.get(DynamicFilterService.class);
+
+    public void storeSummary(String source, TaskId taskId, TupleDomain<String> dynamicTupleDomain)
+    {
+        QueryId queryId = taskId.getQueryId();
+        SourceDescriptor sourceDescriptor = SourceDescriptor.of(queryId, source);
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.computeIfAbsent(sourceDescriptor,
+                ignored -> new DynamicFilterSummaryWithSenders());
+        dynamicFilterSummaryWithSenders.addSummary(taskId, dynamicTupleDomain);
+    }
+
+    /*
+     * registerTasks is required to know whether we have received dynamic filter summaries from all the tasks of query, after which the summary is ready to use
+     * Build side may start reporting summaries through storeOrMergeSummary before registerTasks runs
+     */
+    public void registerTasks(String source, Set<TaskId> taskIds)
```

Comment:
I think you can simply pass `SqlQueryExecution` here. You can get plan from it as well as `QueryInfo`

Reply:
done